### PR TITLE
Use parent to pick proper document during parsing

### LINKED ISSUE #18277: Incorrect Document can be used when creating elements during parsing
Step 1 of https://html.spec.whatwg.org/multipage/#create-an-element-for-the-token states that the intended parent's node document should be used when creating a new element during parsing.

We currently use the document that is associated with the parser. (line 795)
https://github.com/servo/servo/blob/96b4e064a1f0b0fc1ee2811d0c1b528f85a718c6/components/script/dom/servoparser/mod.rs#L787-L804

This causes issues when creating an element that is supposed to be inside a `template`'s `contents` because `contents` has a different document than the `template`.

### PR DESCRIPTION
**Pull request by cbrewster**
Relies on servo/html5ever#312

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18277 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18536)

### CODE REVIEW

**Comment by nox** [comment]
@bors-servo delegate=cynicaldevil

@cynicaldevil Would you be interested in reviewing this?

File: components/script/dom/servoparser/async_html.rs [hunk 1]
```diff
@@ -335,14 +336,14 @@ impl Tokenizer {
                     "Tried to extract contents from non-template element while parsing");
                 self.insert_node(contents, JS::from_ref(template.Content().upcast()));
             }
-            ParseOperation::CreateElement { node, name, attrs, current_line } => {
+            ParseOperation::CreateElement { node, parent, name, attrs, current_line } => {
                 let is = attrs.iter()
                               .find(|attr| attr.name.local.eq_str_ignore_ascii_case("is"))
                               .map(|attr| LocalName::from(&*attr.value));
-
+                let document = parent.and_then(|p| self.nodes.get(&p)).map(|n| document_from_node(&**n)).unwrap_or_else(|| Root::from_ref(&*self.document));
```
Comment (jdm) [inline components/script/dom/servoparser/async_html.rs]: Let's make this a bit more readable with some newlines:
```rust
let document =
    parent
    .and_then(|p| self.nodes.get(&p))
    .map(|n| document_from_node(&**n))
    .unwrap_or_else(|| Root::from_ref(&*self.document));
```

File: Cargo.toml [hunk 2]
```diff
@@ -19,3 +19,5 @@ opt-level = 3
 # crate, add that here. Use the form:
 #
 #     <crate> = { path = "/path/to/local/checkout" }
+html5ever = { path = "../html5ever/html5ever" }
+xml5ever = { path = "../html5ever/xml5ever" }
```
Comment (jdm) [inline Cargo.toml]: These will need to be removed.

**Review by jdm — CHANGES_REQUESTED** [review]
Sorry about the long delay! @cbrewster do you have time to rebase this and address a couple comments?

File: Cargo.toml (hunk 2 above)
Comment (cbrewster) [reply Cargo.toml]: Is it okay to wait for servo/html5ever#312 to merge first?

**Comment by cbrewster** [comment]
I don't have the bandwidth to finish this at the moment.

**Pull request closed without merging**